markdown: Re-enable backslash escape inline pattern.

akshatdalton opened the pull request:
This commit re-enables backslash escape inline pattern.

Fixes #18202.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
New test cases are added.

akshatdalton committed:
markdown: Increase priority of Tex pattern.

This is a prep commit for the issue: #2401.
The Tex pattern is assigned higher priority than Escape
pattern so it doesn't break for cases like $$\{1, 2, 3\}$$.

akshatdalton commented:
The Test cases are added, according to what mentioned in the upstream [`markdown.tests.basic.backslash-escapes.html`](https://github.com/Python-Markdown/markdown/blob/master/tests/basic/backlash-escapes.html).

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2149,6 +2148,11 @@ def build_inlinepatterns(self) -> markdown.util.Registry:
+            markdown.inlinepatterns.EscapeInlineProcessor(markdown.inlinepatterns.ESCAPE_RE, self),
+            "escape",
+            102,
+        )
```
How did you pick 102?  Might be worth commenting on how that's done in python-markdown upstream.

timabbott commented:
So the one thing that I'm worried about with this PR is how it interacts with our LaTeX support, which makes use of a lot of commands like `\int`.  I think it's OK, but just wanted to flag that.

@andersk can you review this one?  I know you've wanted it for a while.

akshatdalton commented:
@timabbott I have added the comment which is similar to the [upstream comment](https://github.com/Python-Markdown/markdown/blob/master/markdown/inlinepatterns.py#L47-L48), the reason why it has to be `102`.

And regarding:
> commands like `\int`

I have added the test cases to what mentioned in https://github.com/Python-Markdown/markdown/blob/master/tests/basic/backlash-escapes.txt which keeps the check of use of `backslash` within a code block and code span. Same I have mentioned in my commit message.

andersk commented:
We should move the `Tex` pattern before `EscapeInlineProcessor`, to avoid breaking the use of backslash in LaTeX such as `$$\{1, 2, 3\}$$` (which we should add as a test case).

We should override `ESCAPED_CHARS` in `class Markdown` to the list of all ASCII punctuation characters as per the [specification](https://spec.commonmark.org/0.29/#backslash-escapes). This matters, e.g., for escaping our mention syntaxes `\@**Cordelia Lear**` and our emoji syntax `\:smile:` (these would also be good test cases).

Have you investigated the difference between `expected_output` and `marked_expected_output`? I guess these are probably just upstream bugs in our ancient version of Marked, which is fine to worry about later.

I picked most of the current priority numbers in #16574 strictly to preserve our prior behavior. We should at some point consider realigning them with upstream’s priority numbers; that can happen separately though.

akshatdalton commented:
@timabbott @andersk 
While investigating, I found that Upstream backslash escaping doesn't work in the case of [autolinks](https://spec.commonmark.org/0.29/#example-306), [raw HTML](https://spec.commonmark.org/0.29/#example-307).
Moreover, it introduced XSS issues, i.e. it failed to escape characters `<`, `>`, `"`, `'` to `&lt;`, `&gt;`, `&quot;`, `&#39;`.

Right now, I'm working on the XSS problem that how can I fix that.

akshatdalton commented:
I have put some comments regarding the XSS issue in[ CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Backslash.20Escape.20Inline.20Pattern).
For the case of [autolinks](https://spec.commonmark.org/0.29/#example-306), [raw HTML](https://spec.commonmark.org/0.29/#example-307), I tried the examples provided there, and it seems that it is disabled in Zulip.

timabbott commented:
Yeah, we've disabled raw HTML because it's a clear XSS path (with autolinks, we have our implementation).  So I don't think we need to worry about the interaction between backslash escaping and raw HTML -- just with the other features that we have enabled.  In any case, I encourage you to proceed with adding tests for cases that would help verify that an implementation doesn't introduce XSS issues or the like as part of a PR like this -- those will be valuable if in the future we look at refactoring the logic, and in my view are 95% of the actual work for this project.

timabbott commented:
This PR will need to also update the documentation.

akshatdalton commented:
I will add the documentation part as the next commit to this PR.

akshatdalton committed:
docs: Document the backslash escape inline pattern.

This commit adds the documentation for the backslash escape inline
pattern, clearly stating how it can be used to generate literal
characters. It is added that Zulip provides backslash escapes for all
the ASCII punctuation characters.

akshatdalton committed:
markdown: Realign Markdown parser priority numbers.

This commit realigns our most of the current Markdown parser's
priority numbers with upstream's priority numbers. These were
kept unchanged in commit 254b90496545c4364e21c78c87720973bf5306f8
strictly to preserve our prior behaviour.

akshatdalton commented:
My 1st commit address the following:

- Added comment why Backtick and Escape have to go before everything else. https://github.com/zulip/zulip/pull/17511#discussion_r594021034
- Added comment: Tex is given higher priority than Escape so that it doesn't break the use of the backslash in LaTeX.
- `ESCAPED_CHARS` list is extended for all the ASCII punctuation characters. https://github.com/zulip/zulip/pull/17511#issuecomment-799872248
- The upstream introduce XSS issues with Backslash Escape. To fix this `UnescapePostprocessor` is added as a subclass to `EscapePostprocessor` (Renaming is done to make it clear what does it actually do). In `EscapePostprocessor` characters are escaped to avoid XSS issues.
- Robust test cases are added to ensure the Backslash Escape pattern doesn't break any of our current features. https://github.com/zulip/zulip/pull/17511#issuecomment-804681217

2nd commit adds the documentation part for the added feature. https://github.com/zulip/zulip/pull/17511#issuecomment-808448247

3rd commit realigns the priority numbers as mentioned by Anders.

@timabbott, @andersk please review.

akshatdalton commented:
Just updated documentation in the `Message formatting` section.

akshatdalton commented:
Just removed the conflicts.

akshatdalton commented:
Just resolved the conflicts!

Key points about the changes are here: https://github.com/zulip/zulip/pull/17511#issuecomment-811072824.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2121,6 +2151,43 @@ def __init__(
+
+        # We have extended the list of Escaped Chars to all the ASCII
+        # punctuation characters to follow the CommonMark.
+        self.ESCAPED_CHARS = [
+            "!",
+            '"',
+            "#",
+            "$",
+            "%",
+            "&",
+            "'",
+            "(",
+            ")",
+            "*",
+            "+",
+            ",",
+            "-",
+            ".",
+            "/",
+            ":",
+            ";",
+            "<",
+            "=",
+            ">",
+            "?",
+            "@",
+            "[",
+            "\\",
+            "]",
+            "^",
+            "_",
+            "`",
+            "{",
+            "|",
+            "}",
+            "~",
+        ]
```
Simpler:

```python
self.ESCAPED_CHARS = list(string.punctuation)
```

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2078,6 +2079,35 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
+    """
+
+    def unescape(self, m: Match[AnyStr]) -> str:
+        return self.char_escape(super().unescape(m))
```
We don’t need a custom escape function.

```suggestion
        return html.escape(super().unescape(m))
```

andersk commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -827,7 +827,8 @@
       "name": "tex_money",
       "input": "Tickets are $5 to $20 for youth, $10-$30 for adults, so we are hoping to bring in $500 from the event ($$x \\approx 500\\$$$)",
-      "expected_output": "<p>Tickets are $5 to $20 for youth, $10-$30 for adults, so we are hoping to bring in $500 from the event (<span class=\"katex\"><span class=\"katex-mathml\"><math xmlns=\"http://www.w3.org/1998/Math/MathML\"><semantics><mrow><mi>x</mi><mo>≈</mo><mn>500</mn><mi mathvariant=\"normal\">$</mi></mrow><annotation encoding=\"application/x-tex\">x \\approx 500\\$</annotation></s ...
+      "expected_output": "<p>Tickets are $5 to $20 for youth, $10-$30 for adults, so we are hoping to bring in $500 from the event (<span class=\"katex\"><span class=\"katex-mathml\"><math xmlns=\"http://www.w3.org/1998/Math/MathML\"><semantics><mrow><mi>x</mi><mo>≈</mo><mn>500</mn><mi mathvariant=\"normal\">$</mi></mrow><annotation encoding=\"application/x-tex\">x \\approx 500$</annotation></sem ...
```
This change seems wrong; this backslash belongs to TeX.

andersk commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -863,7 +864,7 @@
       "name": "tex_error_safe_script_tag",
       "input": "$$\\<script type='text/javascript'>alert('xss');</script>$$\n\n~~~math\n\\<script type='text/javascript'>alert('xss');</script>\n~~~",
-      "expected_output": "<p><span class=\"tex-error\">$$\\&lt;script type='text/javascript'&gt;alert('xss');&lt;/script&gt;$$</span></p>\n<p><span class=\"tex-error\">\\&lt;script type=&#x27;text/javascript&#x27;&gt;alert(&#x27;xss&#x27;);&lt;/script&gt;</span></p>",
+      "expected_output": "<p><span class=\"tex-error\">$$&lt;script type='text/javascript'&gt;alert('xss');&lt;/script&gt;$$</span></p>\n<p><span class=\"tex-error\">\\&lt;script type=&#x27;text/javascript&#x27;&gt;alert(&#x27;xss&#x27;);&lt;/script&gt;</span></p>",
```
This change also seems wrong; this backslash belongs to TeX.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2078,6 +2079,35 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
         return None, None, None
 
 
+class EscapePostprocessor(markdown.postprocessors.UnescapePostprocessor):
```
I think this should not be renamed, as this still performs the operation that upstream calls “unescape”; it just also fixes a bug by adding a missing escape operation.

akshatdalton commented:
Following are the points, I have addressed in my current push:

- Used `list(string.punctuation)`. https://github.com/zulip/zulip/pull/17511#discussion_r624560806
- Used `html.escape` function instead of custom function. https://github.com/zulip/zulip/pull/17511#discussion_r624561450
- Left the name unchanged of `UnescapePostprocessor`. https://github.com/zulip/zulip/pull/17511#discussion_r624564197
- Changes pointed in: https://github.com/zulip/zulip/pull/17511#discussion_r624563720, https://github.com/zulip/zulip/pull/17511#discussion_r624563865 are fixed by https://github.com/zulip/zulip/pull/18342.

@timabbott, @andersk, please review.

akshatdalton commented:
I removed a `TODO`, which I added thinking that there is a bug is upstream.
Here is the issue: https://github.com/Python-Markdown/markdown/issues/1127 that I opened.
I raised the point that `UnescapePostprocessor` should do the HTML escaping when Backslash escaping is done, because for the simple code:

```.py
>>> import markdown
>>> markdown.markdown("{} [] () <>")
'<p>{} [] () &lt;&gt;</p>'
```
they are doing HTML escaping in `<` and `>` characters, therefore, I expected those characters to be HTML escaped when we do Backslash escaping.

They said that since they follow Babelmark and are not a CommonMark implementation, hence the Python-Markdown is more around Markdown's behaviour rather than HTML/browser's behaviour.

To ensure that we follow CommonMark and there is no XSS issue, I've overridden the `UnescapePostprocessor`'s behaviour.

aero31aero commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2220,12 +2220,12 @@ def build_inlinepatterns(self) -> markdown.util.Registry:
-            markdown.inlinepatterns.DoubleTagPattern(STRONG_EM_RE, "strong,em"), "strong_em", 100
+            Tex(r"\B(?<!\$)\$\$(?P<body>[^\n_$](\\\$|[^$\n])*)\$\$(?!\$)\B", self), "tex", 103
         )
-        reg.register(UserMentionPattern(mention.find_mentions, self), "usermention", 95)
```
I don't think this change belongs in the Tex priority change commit.

akshatdalton replied:
I'll do a clean push, resolving all the conflicts.

akshatdalton commented:
@timabbott, a gentle reminder about this PR.
Changes in the current push mentioned in: https://github.com/zulip/zulip/pull/17511#issuecomment-830758087 and https://github.com/zulip/zulip/pull/17511#issuecomment-831653091.

timabbott commented:
@akshatdalton sorry for losing track of this -- do you by any chance have time to rebase this? Or @aero31aero, if you've got time to pick it up.

I'd also be very curious to see a summary of your conclusions on the Python-Markdown response to your issue report.

akshatdalton committed:
markdown: Re-enable backslash escape inline pattern.

This commit re-enables the backslash escape inline pattern. This uses
upstream `EscapeInlineProcessor` and `ESCAPE_RE`.

We have extended the list of `ESCAPED_CHARS` to all the ASCII
punctuation characters, to follow CommonMark.

We also change the frontend `static/third/lib/marked.js` such that it
follows backend logic (mostly).

Fixes #2401.

Signed-off-by: Akshat <akshat25iiit@gmail.com>

akshatdalton commented:
@zulipbot add "buddy review"
@zulipbot add "mentor review"

akshatdalton commented:
@timabbott @andersk can you review this PR? It's been pending for a long time.

sevmonster commented:
Any traction on this? I have vehemently hated the lack of escaping in Zulip ever since I started using it. Of special note is that my group on Zulip commonly uses the Email- or greentext-style `>` strictly as it is textually represented, with no desire to transform it into a quote block. I got around that by adding custom CSS to the web client that renders a `>` before quote blocks, but that doesn't fix the mobile UI. Having to escape `>` is not ideal but it does resolve the issue without any hacky workarounds.

Also, I have been mulling over a way to disable markdown in Zulip for some time, on a per-message basis. After seeing this, I've shifted my thinking to automatically preprocessing user messages if they have a "Disable markdown" box checked to prepend backslashes before non-Zulip markdown (so for example quotes, `@**user**`, `#**stream**`, and attachments will still work). I'd really like to pursue this once this PR gets completed.

Aside, for anyone interested in the markup for the blockquote changes:
```css
.rendered_markdown blockquote {
  color: green;
}
.rendered_markdown:not(.rtl) blockquote {
  border-left: none;
}
.rendered_markdown.rtl blockquote {
  border-right: none;
}
.rendered_markdown:not(.rtl) blockquote:before {
  content: ">";
  float: left;
  position: relative;
  left: -10px;
  width: 0;
  /* from prod blockquote p */
  font-size: 17.5px;
  font-weight: 400;
  /* looks better? */
  line-height: 1;
}
.rendered_markdown:not(.rtl) blockquote:after {
  content: "<";
  float: right;
  position: relative;
  left: 10px;
  width: 0;
  /* from prod blockquote p */
  font-size: 17.5px;
  font-weight: 400;
  /* looks better? */
  line-height: 1;
}
```

alya commented:
#18202 has the most recent discussion on this topic (and links to chat threads).

timabbott commented:
Overall I would be excited to review a correct implementation of this that convincingly does not introduce security bugs.
